samples/nrf9160/modem_trace_flash/src/trace_flash.c

```
@@ -0,0 +1,181 @@
+{
+	int err;
+
+	LOG_DBG("trace_backend_init\n");
+	if (trace_processed_cb == NULL) {
+		return -EFAULT;
+	}
+
+	trace_processed_callback = trace_processed_cb;
+
+	if (magic != magic_val) {
+		LOG_DBG("Initializing .noinit\n");
+		magic = magic_val;
+		traces_size = 0;
+	}
+
+	traces_size = 0;
+
+	err = flash_area_open(FLASH_AREA_ID(modem_trace), &modem_trace_area);
+	if (err) {
+		LOG_ERR("flash_area_open error:  %d", err);
+		return -ENODEV;
+	}
+
+	err = flash_area_has_driver(modem_trace_area);
+	if (err != 1) {
+		LOG_ERR("flash_area_has_driver error: %d\n", err);
+	}
+
+	flash_dev = flash_area_get_device(modem_trace_area);
+	if (flash_dev == NULL) {
+		LOG_ERR("Failed to get flash device\n");
+		return -ENODEV;
+	}
+
+	err = stream_flash_init(&stream, flash_dev, stream_buf, BUF_LEN, TRACE_OFFSET, TRACE_SIZE,
+		NULL);
+	if (err) {
+		LOG_ERR("stream_flash_init error: %d", err);
+		return err;
+	}
+
+	// erase flash partition first to write without interruptions while tracing
+	int64_t erase_before = k_uptime_get();
+
+	err = flash_erase(flash_dev, TRACE_OFFSET, TRACE_SIZE);
+	if (err) {
+		LOG_ERR("flash_erase error: %d", err);
+		return err;
+	}
+	int64_t erase_time = k_uptime_delta(&erase_before);
+
+	LOG_DBG("Erased %lu bytes in %lld ms", (unsigned long) TRACE_SIZE, erase_time);
+	LOG_DBG("Erase speed %lld kb/s ", (TRACE_SIZE * 8 / erase_time));
+
+	tot_write_time = 0;
+
+	LOG_DBG("Modem trace flash backend initialized\n");
+
+	return 0;
+}
+
+int trace_backend_deinit(void)
+{
+	LOG_DBG("Flash trace backend deinitialized\n");
+
+	return 0;
+}
+
+static bool has_flushed;
+
+int trace_backend_write(const void *data, size_t len)
+{
+	if (has_flushed) {
+		trace_processed_callback(len);
+		return -EIO;
+	}
+
+	int ret = internal_flash_write(data, len, false);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	trace_processed_callback(len);
+	if (ret < 0) {
+		LOG_ERR("Unexpected long wait for flush mutex");
+		return ret;
+	}
```

```suggestion
	ret = trace_processed_callback(len);
	if (err< 0) {
		LOG_ERR("trace_processed_callback returns %d", ret);
		return ret;
	}
```

---

Thanks, the code here was not correct, but we do need to return an error when the callback returns an error. That means we called the callback with an invalid length, so we should return an error and abort the modem tracing (which nrf_modem_lib_trace handles).